### ISSUE TITLE
Failing push notification

### ISSUE THREAD

**Issue by ageru** [ChatSecure/ChatSecure-iOS#770]
Hi,

Sorry to create yet another bug about push...

**Setup**
- Server: prosody .9.12 with CA-signed TLS certificate
- Relevant modules: cloud_notify, smacks, pinger, carbons, mam, csi, filter_chatstates, throttle_presence.
- IPhone SE with latest iOS. Low battery mode off, all is ON in General Settings > ChatSecure.
- Chatsecure 4.0.8

**Steps to reproduce**
1. Install Chatsecure, log to server
2. Send a couple of messages to user on iPhone, they reach him, OMEMO works... Everything looks clear.
3. Confirmed push is active in the app's interface
4. Leave the app.

**Actual result**
- Quickly, the iPhone user shows as away on other devices, then as offline.
- Messages don't reach the device unless the iPhone's user starts the app
- The iPhone's user has to specifically open the app to get the sent messages
- Relevant portions of the server log:
  - Successfully initiating push (it seems):
`May 14 HH:MM:SS [alphanum characters]     info    Push notifications enabled for [ID]@[chatdomain.tld]/chatsecure (pubsub.chatsecure.org<A3486BB3-6442-4945-84E2-[more alphanum characters])`

  - When trying to send a message to offline device (the last number increments with every try of the server):
`May 14 HH:MM:SS [chatdomain.tld]:cloud_notify    info    Got error of type 'cancel' (not-allowed) for identifier 'pubsub.chatsecure.org<A3486BB3-6442-4945-84E2-[alphanum characters]': error count for this identifier is now at 1`

**Expected Result**
If iPhone is connected to the Internet, the iPhone user gets a notification straight away.

**Bonus questions**
- What does mod_pinger do? I can make an educated guess looking at the name and bits of its code, but I'm not 100% certain.
- Would it be possible to get a wiki of FAQ detailing exactly what modules in prosody are needed to use all the features of ChatSecure? @chrisballinger gives a few times his config in threads here and there, but that's hardly discoverable and probably not comprehensive or up to date.

Many thanks.

**Comment by chrisballinger** [comment]
Thanks for the thorough investigation. I've never seen that error message before from the push module. It looks like the code has been updated recently: https://hg.prosody.im/prosody-modules/file/tip/mod_cloud_notify/mod_cloud_notify.lua#l73

Perhaps try rolling back mod_cloud_notify to a (much older) commit?

**Comment by KingKili** [comment]
I have the same issue with https://trashserver.net/

**Comment by ageru** [comment]
All the modules are automatically updated weekly on my server (and I didn't have any ChatSecure user before the past week-end) so that would be consistent with your theory. As a temporary fix, rolling back to an earlier commit should be possible, I just hope it won't break things with the users running Conversations.

Would you be able to let me know what commit of this module you are using, or alternatively to date it?

By the way, when I noticed the issue, I installed mod_pinger to try to fix it, since you seemed to be using it. Is it in any way necessary to have it installed for ChatSecure and Push to work properly?

**Comment by chrisballinger** [comment]
mod_pinger shouldn't be required. Not sure the exact mod_cloud_notify commit, but I'll figure that out and test against the latest commit. Push is still working fine as of a few minutes ago on my personal server configuration.

**Comment by paskalito** [comment]
could this be avoided using this push mechanism ?
https://github.com/ChatSecure/ChatSecure-iOS/issues/650#issuecomment-289429776

**Comment by beeloo42** [comment]
I'm also encountering this issue
I tried the 5 latest `mod_cloud_notify` versions without success...

**Comment by afriedmanGlacier** [comment]
We are using eJabber and have changed a bit of the underlying functionality of push, but just thought I'd chime in and say push still works ok for us. However...

@chrisballinger the one problem I have seen lately, I thought I'd add here just in case it is relevant to these issues is that our RubDub has died unexpectedly a few times. Haven't figured out why yet (and we also haven't updated to the RubDub with Rollbar yet). When this happened, eJabber could not connect to it and push stopped working. Have you seen this at all Chris? I'm not sure if you guys use the same RubDub and ChatSecure-Push-Server setup as is available on github?

If you have not had this problem, we'll just monitor it on our end, but just thought I'd add this here in case it is relevant.

**Comment by chrisballinger** [comment]
Yeah it does sometimes, but I re-provisioned our server don't see this anymore. I have an external port monitor that emails me when it goes down, and it hasn't had an issue since then. (Ubuntu 16.04 and latest Node PPA.)

**Comment by therob84** [comment]
Exactly the same problem I encounter with two of my contacts who use ChatSecure 2.0.8 (while I am using Conversations 1.18.4). All of us are using jabber.de (prosody >0.9.10)

- At iOS devices Push is active and running (as ChatSecure me confirmed).
- For about 1-2 days delivering messages to the iOS-devices works perfectly:
  - I compose a message (Conversations) while iOS-users are displayed as offline
  - triggered by the typing-notifier iOS-users are shown as "online" for some seconds
  - i send the message some time later, iOS again return to "online", receives the message, send an delivery-notifier - perfect!
- But after 2-3 days this does not work anymore and the iOS user has to reopen ChatSecure to get Push activated again (which is not very userful for long-term conversations....), otherwise the message never will be delivered.

I think it is the same problem than stated above, but I could not provide any logfile but are eager to help solving this problem. 

Is the reason now already undrestood or is it still a mystery, why/when/where this happens?

I hope the best for a fix in the next release....cheers, Robert

**Comment by chrisballinger** [comment]
If the Prosody server restarts it will lose all the push tokens because
they are only stored in memory. Is the server restarting on regular
intervals?

**Comment by therob84** [comment]
Hm...I have to ask the guys from jabber.de. But I doubt that they would
do it that often...
is there a workaround/solution if this is the problem?
Is there any other idea to improve this?

**Comment by chrisballinger** [comment]
mod_cloud_notify should store push tokens in permanent storage so they
persist across restarts

**Comment by therob84** [comment]
@chrisballinger 
Anwser from jabber.de-guys are: 
- "Server is restarted about once a month" ....so I don't think this is the explanation for me (but can't say for sure)
- they "can't find the ability for storing a session" (I guess in mod_cloud_notify)
- but they updated today the module Cloud_Notify (don't know between which versions) and I will check if it is improved (should the ChatSecure-Apps be restarted or should they use the new modul version automatically in background?)

Any more suggestion about the problem from your side?
For my information: is anywhere written how the ChatSecure-Push-Servers are working together with the iOS devices and with the jabber-servers? I can't understand how it works in principle until now...

**Comment by therob84** [comment]
Two more questions I want to raise:
- how does ChatSecures push-functionality behaves, after CS is activley "swiped out" in taskmanager or even forced killed?
- Or if iOS is freshly restarted an ChatSecure is not yet launched manually by user interaction?

--> Will CS (re-)establish a Push-Session in this scenarios or is there first an user activity needed (which would not be that relieable and practicable)?

**Comment by chrisballinger** [comment]
Both of those scenarios disable push until the app relauches due to
limitations of the `content-available` pushes we use. There is discussion
about alternative approaches but the development budget is very thin at the
moment.

**Comment by mimi89999** [comment]
Zom-iOS has the same issues, right?

**Comment by chrisballinger** [comment]
yes

**Comment by mimi89999** [comment]
But how does Facebook, Twitter, Snapchat, etc. work then? They don't do VoIP and users are still able to get notifications after swiping the app and after a [re]boot.

**Comment by chrisballinger** [comment]
They send your messages in plain text to Apple, which is unacceptable for
an app like this.

**Comment by mimi89999** [comment]
Can't we send *you have unread messages. Please open the app to see them* or something similar then?

**Comment by chrisballinger** [comment]
It would require some additional server side logic / filtering because
these events are sent very frequently, for instance, all typing
notifications. Once the push event reaches our pubsub node we don't have
enough information on whether or not to filter it. There was some
discussion on extending this format, in some of the previous issues. I'm
not saying the current design is perfect, but funding is tighter and the
dev team is smaller, so we really have to focus on one thing at a time.

**Comment by mimi89999** [comment]
@chrisballinger It would be a huge UX improvement. Users won't need to remember not to swipe the app from recent or to launch it after [re]boot.

As for presence, status, etc. smacks does that very well. We could also check for the `<body>` tag in the stanza in the cloud_notify module and send a push message only if there is one.

**Comment by mimi89999** [comment]
@chrisballinger Will it require a lot of work to implement those high priority push notifications in the app and on the node?

**Comment by therob84** [comment]
I just edited my [comment above](https://github.com/ChatSecure/ChatSecure-iOS/issues/770#issuecomment-303657233), about only updating the Cloud_Notify module on the prosody-server (of jabber.de) can **NOT** solve the problem that ChatSecure/push-service is not restarted automatically after "restartingOS/swiping CS away/kill CS"... And thus is a very deadly drawback and makes CS nearly unusable by iOS-contacts which do not care that much about "always checking if CS is launched and running" ......... :-(

**Comment by chrisballinger** [comment]
@therob84 I have plans to improve this in an upcoming release. Going to take the approach of sending generic foreground "New Message" notifications and possibly "Someone is typing..." as well.

**Comment by paskalito** [comment]
@chrisballinger 
Hi that's sounds so great to me!! thank you!
i wanted to post a bounty on this, but i somehow could not find the issue there.. - did you stop supporting bountysource? ( looks like the last push bounty didn't end well..)
btw. is the project on gratipay (couldn't find it) or liberpay?

**Comment by chrisballinger** [comment]
@paskalito You can now set up a [monthly donation](https://chatsecure.org/blog/sustainable-open-source-starts-with-you/) via in-app purchase within the app. I also have PayPal and Bitcoin options if you'd like to [give a little extra](https://github.com/ChatSecure/ChatSecure-iOS#cost-for-end-users):

Thanks for your support!

**Comment by therob84** [comment]
@chrisballinger: I saw your announcment for beta-testing 4.1.0 at twitter (with http-upload and inline images) - great !!!!!!
Could you also include already some improvments about behaviour of push messages (after restart,kill-app etc.) in this release?
Great work, chris!

**Comment by chrisballinger** [comment]
Those improvements are targeted for the v4.1.1 release

**Comment by chrisballinger** [comment]
I have some unfortunate news. Without changes to the XEP-0357 spec and server implementations to send a `message_type` (`typing`, `message`, etc) to my pubsub node, this approach will not be feasible. I need to distinguish between typing notifications and messages with a body, otherwise you end up with a nightmare scenario where you get a constant flood of "New Message" notifications that don't correspond to actual messages AND can no longer receive the actual messages because the app does not reliably wake up in the background.

Even if I can filter out non-messages, it looks like the app is no longer waking in the background from silent pushes. Without those background wakeups, you won't be able to see any actual messages. I am wondering if the foreground pushes are somehow blocking the background content fetch from coming through.

It looks like the only way to really make this work reliably will be to add WebRTC for VoIP PushKit support.

Sorry everyone.

![screen shot 2017-06-27 at 4 50 04 pm](https://user-images.githubusercontent.com/518687/27614902-55ed45de-5b59-11e7-9612-6128bf2aff2a.png)

**Comment by afriedmanGlacier** [comment]
@chrisballinger I am wondering if there might be other options. Ours works as one would expect, and without the spamminess, though I am pretty sure we do broke some of the XEP specifications in the process. We have control of the server, and some of our changes are server side, so maybe its not possible, but maybe worth talking through what we did to see if there are any other possible routes. 

Here are kind of the basics of what we did if I remember correctly:
1) On the server, we use @weiss version of mod_push for eJabber, and modifed it to remove some of the notifications that aren't actual messages and we didn't need or want
2) We log out of ChatSecure whenever the app goes into the background (I do realize this could be a problem with MUC, but we were waiting until the upcoming MUC fixes to see what needs to happen there). I think Signal does this too.
3) Because we log out in the background, we can filter out some typing and other notifications based on whether a "buddy" is online or not. 
4) We use the "New Message" non-silent approach and only get push messages when an actual message is sent.

I don't have time for details tonight, but probably tomorrow if you are interested. Let me know

**Comment by chrisballinger** [comment]
Yeah the main problem is that, by design, I don't control who is sending me what. There's really no way for me to enable this without horribly breaking the UX for many people. Even if I can filter out the typing notifications... I can't even seem to use content-fetch anymore when also sending foreground notifications, so you'll never see the message content until you open the app, which is even worse imo.

**Comment by mimi89999** [comment]
@chrisballinger The server module could be easily updated to only send a message to the pubsub node if there is a body. It could also advertise the version, so that you could know if it is the updated module or not.

I saw your screenshot. Did you write code for that? If yes, where could I find it (or even better the assembled app since I can't really build for iOS on GNU/Linux)?

**Comment by afriedmanGlacier** [comment]
@chrisballinger ahhhhh, I forgot, we also don't do the fetch. Because we are receiving "New message" notification for every actual message, we decided we'd rather just open the application to see the message and have Push notifications working the way we'd expect than to worry about seeing the content in the notification. For us, this seems better than the alternative.

**Comment by chrisballinger** [comment]
@mimi89999 This technically doesn't require any client changes. Only on the server.. but the server only works for the App Store build of the app.

@afriedmanGlacier Yeah...

**Comment by mimi89999** [comment]
So I won't be able to test anything :-(

I believe that changing the server module and having it send the version would solve the issue.

**Comment by mimi89999** [comment]
And *New Message* is better than no notification at all.

**Comment by chrisballinger** [comment]
Hm yeah I guess I could make this behavior opt-in via a mod_cloud_notify patch and some experimental XEP-0357 spec extension for specifying the message type. I think it might still work with content fetch because my iOS 11 test device cannot seem to do silent push / content fetch at all anymore, so there must be something else going on as well.

**Comment by therob84** [comment]
(Is there a connection with this topic?)
A friend of mine cannot connect to CS-Push-Service (red alert-sign in settings) with v4.1. He has the iOS "Background connection" disabled (for battery reasons) - but claims that threema and whatsapp still work in delivery of push messages. With CS there is none push message delivered until now. The "reset" button in push menu brought no solution. A screenshot I will upload in a few minutes.

Do you know, where the error could be located and how to solve?

**Comment by hemee** [comment]
@therob84 He has to switch on the background refresh. Pushes are rely on it.

**Comment by therob84** [comment]
@hemee 
- is this the only reason/solution? 
- Can it only be allowed for CS? 
- If it would be switched on, how often CS will do anything in background? 
- Is it a battery eater or should there be no difference visible?
- Why WhatsApp and Threema do work properly without switching it on?

My problem is that I have 1.) successfully convinced a friend of using xmpp/ChatSecure 2.) Now I have to tell him, that he should enable the background refresh for using it, which he costs battery, although WhatsApp and threeme don't need it (as it seems)....... bad chain of convincement unfortunately.... :-(

**Comment by ageru** [comment]
- Probably is
- Yes, it's app specific: go to Settings, scroll down to ChatSecure, enable Background data, done! No other app's setting is modified 
- Not more than before. If I understand correctly, it should be just enable the app to receive push messages when it's not in the foreground in this case. 
- Not sure about this one

You are only activating Background data for ChatSecure, and that's only to allow it to receive stuff, not to send continuously.

**Comment by therob84** [comment]
- maybe @chrisballinger can say anything about battery drain between background refresh ON/OFF for CS and why other programs can use push without background refresh enabled?

**Comment by STPKITT** [comment]
I'm not into Apple, but doesn't Apple provide a push-network like Google does with it's GCM? I always thought that was the case and that WhatsApp and similar apps use that push-service and therefor don't have the problems ChatSecure has to deal with.

**Comment by chrisballinger** [comment]
High priority notifications require either:

1. Sending plaintext to Apple
2. Supporting VoIP and abusing the API intended for call notifications

Right now neither of these are feasible solutions. I have been experimenting with sending generic "New Message!" pushes, but it seems to break the content fetch.

**Comment by paskalito** [comment]
@money : thank you, as i don't own anything made by apple, and this as an expression of a strong belive in free culture etc. thanks for still providing other donating options i think it's important!
also you may want to have a look into http://gratipay.com/ 

@push : thanks for starting to work on that! sad it doesnt work with push AND content fetch. but i feel VERY stongly in the direction of @afriedmanGlacier and @mimi89999 that it should be a absolute priority to get any reliable push. and content fetch could be a nice extra. 
so if there is anything to opt in as a user (if i understood correctly that that was to discuss) IMHO it should be the content fetch function with a big warning that it will probably break push functionality.

so if it's possible the get push working with "new message" notification only. but reliably with any server that fullfilsh the push XEP. I beg you to integrate that into the next version of cs. and make that behavior default.

because that will all the ios users i know to actually start using the app. 
my experience with cs is i write to all my ios contacts and i never get any reply. - because they also never see the messages...

**Comment by STPKITT** [comment]
Why isn't push via VoIP API with PushKit a feasible solution when that's what WhatsApp uses?

**Comment by stigger** [comment]
Because WhatsApp supports VoIP calls and ChatSecure does not? You cannot just abuse the API, it won't get past Apple's review.

**Comment by STPKITT** [comment]
Ah that makes sense, I didn't consider Apple's policies.

**Comment by stigger** [comment]
By the way, from my experience the content-available push notifications are quite reliable. I've had other issues that indirectly affected the reliability (some push-related issues on the server-side, long processing of pushes due to #812, losing incoming messages due to #813), but once I resolved them, everything started working pretty smoothly.

So, it could be that the VoIP API is not really necessary after all. I suggest waiting until the above issues are resolved in master and checking if it helped.

**Comment by mimi89999** [comment]
@chrisballinger Any news on implementing what I proposed?

**Comment by chrisballinger** [comment]
Yes I've made some progress, but I'm still catching up after returning from
a break

**Comment by mimi89999** [comment]
@chrisballinger I applied your changes to `mod_cloud_notify` from https://groups.google.com/forum/#!topic/prosody-dev/Ge3USXmuMIc and after resolving an issue with connecting to the pubsub node (#822) I tested it.

I am very happy with how it works. When I send a message and CS is not running in the background, I get almost instantly the *New Message!* notification on my iPad with another account. It works even after rebooting the device. Thanks, to the priority, I don't get flooded with notifications. I get only one notification per message when I send the message. I haven't tested it in real conditions yet tho, but I expect it to work well since it worked now.

**Comment by chrisballinger** [comment]
@mimi89999 Aww yeah! Glad you like it.

**Comment by MattRiddell** [comment]
Yep, +1 from me. Same thing, with that patch to mod_cloud_notify it works fine.

I had some problems with OTR not working after resumption (Adium says: The following message was not encrypted: OTR Error: Encryption Error), but OMEMO seems to work fine.

**Comment by chrisballinger** [comment]
@MattRiddell That's because OTR sucks on mobile and there's really nothing we can do to fix it other than move everyone over to OMEMO. :)

**Comment by mightyBroccoli** [comment]
I am using prosody 0.10 nightly and some of my user notably these form iphone are not able to get push notifications.

Mostly they get token after token and the app will always say cloud push is not possible.
@chrisballinger Are there any steps I need to complete as a hoster to fully support push for iphone users?

**Comment by chrisballinger** [comment]
@mightyBroccoli Are you running `mod_cloud_notify`?

**Comment by ageru** [comment]
I've lost my iPhone users as well for the past few days, but noticed it today. I'm still on Prosody 0.9 though. 

In the token interface of ChatSecure (4.1.0), they get something along the lines of 'this server doesn't support the functionality'.

Could it be something with the token server? I'm asking because mod_cloud_notify hasn't been updated for weeks, so that doesn't leave much else that could have changed and provoked the issue.

**Comment by MattRiddell** [comment]
Just a heads up Chris - I'm using your custom mod_notify from the google group and am receiving notifications on my iphone fine.

**Comment by chrisballinger** [comment]
Ok thats good news. Debugging other people's server configurations is
probably my least favorite part of developing this app. Not very much
positive feedback these days. Sigh

**Comment by mightyBroccoli** [comment]
@chrisballinger yes I am using mod_cloud_notify. Is there anything I can do to help iPhone users use push?

**Comment by mimi89999** [comment]
@mightyBroccoli Be sure that mod_cloud_notify works correctly and that you are using the latest versions of modules... Also, https://bitbucket.org/chrisballinger/prosody-modules/branches/compare/12b4387..b1c5b0c#diff should greatly improve experience. It works very well for me (or at least it did when I tried several days ago).

@chrisballinger You probably can't do much about server setup... What were the feedbacks?

**Comment by mightyBroccoli** [comment]
@chrisballinger i will try your cloud_push changes. I think if they really work that well why don't you push them to the prosody community modules hg?

**Comment by chrisballinger** [comment]
@nico because it involves changes to the xep-0357 spec, and i still have
some other ideas to improve it before standardization

**Comment by Echolon** [comment]
@chrisballinger @MattRiddell Hey, my provider (jabber.de) has this mod_notify activated for a while already. Doesn't work for me so far - sorry. Their apps will be set to doye or however it works on iOS :(
(Conversations to ChatSecure, both at jabber.de)

**Comment by mimi89999** [comment]
@Echolon But they probably don't have the changes...

**Comment by MattRiddell** [comment]
Yep exactly. 

I actually was trying to get everything working with ejabberd but saw this thread and did a 100% fresh install of prosody from source with that new version on Debian 8.0 and it worked flawlessly.

**Comment by Echolon** [comment]
Okay, thank you. I'm going to check this.

**Comment by ageru** [comment]
Sorted my latest problem:
- applied Google Group/BitBucket patch to mod_cloud_notify 
- relaunched prosody
- added hostname to chatsecure config

And my iOS users seem to be back in business! Problem seem to have started after I renewed the Let's Encrypt cert for the domain. I found pubsub and cloud_notify errors that seem related in my logs, I can post them if anyone's interested. 

Thanks!

**Comment by Wiuma** [comment]
Modified mod_cloud_notify ( BitBucket) for prosody  and push works like a charm.

**Comment by Echolon** [comment]
www.jabber.de seems to work now as they updated mod_cloud_notify

**Comment by therob84** [comment]
> www.jabber.de seems to work now as they updated mod_cloud_notify
Could be the result that I sent them yesterday an e-mail and they instantly updated the package for their prosody server.

But my question is: what "seems to work" now exactly better than before? What is now different than 3 days before and how did you check this and is it dependend on CS version?

**Comment by Echolon** [comment]
@therob84 https://twitter.com/jabberde/status/898584861316833282

They said they updated the unofficial changes from mod_cloud_notify. Before they only ran prosody stable as I understood. What works better? My iOS ChatSecure contacts now receives push. before they didn't. All current CS version.
(I'm afraid one guy now receives some kinda ghost push notifications, but thats better than nothing :D)

**Comment by mimi89999** [comment]
@chrisballinger When the app isn't running, the "New message" push messages show on my devices, but there isn't any sound. If I am not using the device when it comes, I don't know about it.

**Comment by ageru** [comment]
I'm like you @mimi89999 : my iPhone users are there but, they still don't receive push :( Not sure what I am doing wrong...

Prosody tells me this, which sound good:

`Aug 19 15:29:04 c2s41b621c0     info    Push notifications enabled for user@domain.tld/chatsecure (pubsub.chatsecure.org<A3486BB3-6442-4945-84E2-7736229C8CF5)`

But then I get this every so often:

`Aug 19 16:14:20 [domain.tld]:cloud_notify    info    Got error of type 'cancel' (not-allowed) for identifier 'pubsub.chatsecure.org<A3486BB3-6442-4945-84E2-7736229C8CF5': error count for this identifier is now at 3
`
And sometimes it all seems to all go awry:

```
Aug 17 10:47:52 circle-area.net:cloud_notify    info    Got error of type 'cancel' (not-allowed) for identifier 'pubsub.chatsecure.org<0584D5A1-9939-484C-A92A-A4E5B4EF0560': error count for this identifier is now at 85
Aug 17 10:47:52 circle-area.net:cloud_notify    warn    Disabling push notifications for identifier 'pubsub.chatsecure.org<0584D5A1-9939-484C-A92A-A4E5B4EF0560'
Aug 17 10:47:52 mod_c2s error   Traceback[c2s]: ...er/modulesenabled/mod_cloud_notify/mod_cloud_notify.lua:80: attempt to index a nil value 
stack traceback:
        ...er/modules-enabled/mod_cloud_notify/mod_cloud_notify.lua:80: in function <...er/modules-enabled/mod_cloud_notify/mod_cloud_notify.lua:62>
        /usr/lib64/prosody/util/events.lua:67: in function </usr/lib64/prosody/util/events.lua:63>
        /usr/lib64/prosody/util/events.lua:67: in function 'fire_event'
        /usr/lib64/prosody/core/stanza_router.lua:187: in function 'core_post_stanza'
        /usr/lib64/prosody/core/stanza_router.lua:205: in function 'core_route_stanza'
        /usr/lib64/prosody/core/stanza_router.lua:219: in function 'core_route_stanza'
        /usr/lib64/prosody/core/stanza_router.lua:191: in function 'send'
        ...er/modules-enabled/mod_cloud_notify/mod_cloud_notify.lua:268: in function 'handle_notify_request'
        ...er/modules-enabled/mod_cloud_notify/mod_cloud_notify.lua:286: in function <...er/modules-enabled/mod_cloud_notify/mod_cloud_notify.lua:284>
        /usr/lib64/prosody/util/events.lua:67: in function 'fire_event'
        ...
        /usr/lib64/prosody/util/xmppstream.lua:255: in function 'feed'
        /usr/lib64/prosody/modules/mod_c2s.lua:233: in function 'data'
        /usr/lib64/prosody/modules/mod_c2s.lua:255: in function 'onincoming'
        /usr/lib64/prosody/net/server_event.lua:636: in function </usr/lib64/prosody/net/server_event.lua:582>
        [C]: in function 'loop'
        /usr/lib64/prosody/net/server_event.lua:797: in function </usr/lib64/prosody/net/server_event.lua:796>
        [C]: in function 'xpcall'
        /usr/lib64/prosody/../../bin/prosody:376: in function 'loop'
        /usr/lib64/prosody/../../bin/prosody:407: in main chunk
        [C]: at 0x4ea4f8e690
```

I double-check the patch I applied to cloud-notify, but I don't see any mistake.

Any one has a clue?

**Comment by mimi89999** [comment]
The errors you get are most likely because the token has expired and you haven't got a new one.

As for that error with the trace:
It looks like there is a bug in the code handling errors from CS pubsub push server.

**Comment by ageru** [comment]
Thanks for your insight!

The token errors are strange: checking the token interface in CS, I never see it run out. There are always a few unused token.

I haven't pinned down yet when the trace error happens. I'll try to see with my users if it's linked to coming back online, picture transfer, or something else.

**Comment by mimi89999** [comment]
@ageru It is exactly here: https://bitbucket.org/chrisballinger/prosody-modules/src/12b4387c0991eb24a15677b8bf59184ed880fa40/mod_cloud_notify/mod_cloud_notify.lua?fileviewer=file-view-default#mod_cloud_notify.lua-80

Just after `module:log("warn", "Disabling push notifications for identifier '%s'", push_identifier);`  https://bitbucket.org/chrisballinger/prosody-modules/src/12b4387c0991eb24a15677b8bf59184ed880fa40/mod_cloud_notify/mod_cloud_notify.lua?fileviewer=file-view-default#mod_cloud_notify.lua-78

I never got such a token error (or am not aware of it)

**Comment by tmolitor-stud-tu** [comment]
The backtrace was caused by a bug I fixed a few minutes ago.
You can get the latest version of mod_cloud_notify here: https://modules.prosody.im/mod_cloud_notify.html

**Comment by ageru** [comment]
Many thanks for this @tmolitor-stud-tu and @mimi89999 ! I have pulled and up-ed the change from Mercurial in the last hour. No more backtrace so far, I'll keep an eye on my logs.

Following that, I have re-implemented the "Google Group" changes manually in that updated mod_cloud_notify.lua fill. Still no push notification for iPhone users, and I still get the "cancel" error. I had a closer look at the file, and mine seem very different from the "Google Group" one. Chris did remark that he is working from an old version of Cloud Notify, which must be way too different from mine for the patch to actually do its deed. Oh well.

**Comment by mimi89999** [comment]
@ageru Try that: https://lebihan.pl/files/mod_cloud_notify.lua I haven't tested it yet, but should be OK...

**Comment by Loader23** [comment]
The file is working, but still no sound when the app isn't running :-(

**Comment by ageru** [comment]
@mimi89999 That's very kind of you to have taken the time, thank you :)

 I have been using your file for a few hours now. I don't have my test iPhone at hand, but from what my iOS users tell me, they still don't get push notification. And the "cancel" message still shows up regularly in my Prosody logs. But otherwise it doesn't seem to hurt. I'll try to test this more thoroughly in the coming days, see if I can find out something more.

**Comment by mimi89999** [comment]
@Loader23 I also don't have any sound for the notifications. As for the other issue, are you using any e2e encryption?

@ageru Maybe your users should try to register again with push? Under server info, under push tokens they should choose reset.

**Comment by chrisballinger** [comment]
I see why people aren't getting sounds but still are receiving "New
Message!" notifications. Should be an easy fix without requiring a client
update, hopefully can push out tomorrow if have time.

**Comment by Loader23** [comment]
@mimi89999  Iam using OMEMO. All Fingerprints are accepted.  I guess the time between the Client change was too short. But I have ignore_presence_priority = true and carbons in my config for this. 

@chrisballinger That sounds good. When are you going to release 4.1.1 for iOS? Iam also affected by [https://github.com/ChatSecure/ChatSecure-iOS/issues/777](https://github.com/ChatSecure/ChatSecure-iOS/issues/777 ) and an update would be nice.

**Comment by chrisballinger** [comment]
@Loader23 Probably by week of Sep 10-16, beta week before

**Comment by Loader23** [comment]
@chrisballinger Great :-) 
Please give short Feedback if you have fixed the notification Sound bug ;-)

**Comment by ageru** [comment]
Got finally the possibility to test properly, apologies for the delay.

@mimi89999 : the backtrace has not shown up since I applied your fix. Thanks :)

Client side, I reset Push, and the client seem to receive notifications when the device's screen is turned on and unlocked. Is that the expected behaviour at this stage, or should the device receive notification also when the device is locked/screen turned off?

Server side, I still get the "cancel" error, possibly for every message sent or receive through Chatsecure, considering how much of those I see.
Also, Chatsecure seems to registers to receive push notification every time the app is opened by the user. It shows as this fairly often in the logs as this:

```
Aug 26 22:27:36 c2s41ee6888     info    Authenticated as user@domain.tld
Aug 26 22:27:36 c2s41ee6888     info    Push notifications enabled for user@domain.tld/chatsecure (pubsub.chatsecure.org<A3486BB3-6442-4945-84E2-7776229C5BF5)
```

Again, not sure if all this is expected behaviour or not.

I'll keep looking for hints in the coming days.

**Comment by danielreuther** [comment]
As I understand from the discussions, there is currently a working version of mod_cloud_notify for Prosody. And according to https://github.com/ChatSecure/ChatSecure-iOS/issues/637#issuecomment-287409197, I'm assuming that @weiss may start working on a similar solution once it things have been standardized.

If my understanding is correct, is there anything I can do in the meantime? I'm running the mod_push version from @weiss's repository.

**Comment by Loader23** [comment]
@mimi89999 I found out, that carbon message on iPad is working fine, as long as the app is open in the background. Is this intended?

@danielreuther With the mod_cloud_notify from mimi89999 push is working fine with app in background.

**Comment by danielreuther** [comment]
@Loader23 Right, but it's a Prosody module. Hence my question about ejabberd -- I'm currently assuming that mod_push would require similar modifications and I'm afraid those won't happen before the adjustments found their way into the spec.

**Comment by mimi89999** [comment]
@chrisballinger Did you solve the issue with the notifications being silent>

**Comment by Loader23** [comment]
@chrisballinger or could you tell us the year of the "tomorrow" you are going to fix it?

**Comment by tmolitor-stud-tu** [comment]
@Loader23 this is just unfair. Push on iOS is very complicated and frankly: if you want a platform that doesn't make life so hard for its developers just switch to Android.

**Comment by Loader23** [comment]
@tmolitor-stud-tu That has nothing to do with being unfair. It is ok if he did not have the time to fix it yet, but if he says "tomorrow" some feedback of how its going would have been nice after more than two weeks.

**Comment by Wiuma** [comment]
"New Message" info works for me with sound since 2 days.

THX @chrisballinger

**Comment by weiss** [comment]
@Loader23, maybe just cut down your payment for his development work next time he fails to meet deadlines?

**Comment by Loader23** [comment]
@weiss You don't get it either, maybe you just read what i wrote again until you understand it ;-)

**Comment by weiss** [comment]
Or maybe you don't get it?  I can't speak for @chrisballinger, but as far as I'm concerned, this sort of nagging is one of the most annoying things when working on open source projects.

**Comment by Echolon** [comment]
@weiss Yes - you are so I right - I now should definitely reclaim my 12$ donation! ;)

**Comment by chrisballinger** [comment]
I fixed the notification sound a few days ago but forgot which issue # was tracking it.

**Comment by mimi89999** [comment]
@chrisballinger I have the notification sound now. Thanks a lot.

@Loader23 Are you still experiencing those issues? What ChatSecure and iOS version do you have? Does your server have the cloud_notify experimental module?

**Comment by Loader23** [comment]
@chrisballinger No Problem, thanks alot for fixing it.

@mimi89999 I have the mod_cloud_notify.lua installed you posted above. Maybe its normal to not receive any Messages when the app is closed and the message is already received by Conversations?

**Comment by thundergreen** [comment]
Indeed @Loader23 as Chatsecure has no MAM

2017-09-11 8:39 GMT+02:00 Loader23 <notifications@github.com>:

**Comment by Loader23** [comment]
@thundergreen Yep, thats explains it. Thanks for info.

**Comment by weiss** [comment]
@danielreuther:
> I'm currently assuming that [ejabberd's] `mod_push` would require similar modifications and I'm afraid those won't happen before the adjustments found their way into the spec.

I think we could cheat a bit and make it work by setting the [existing][1] `last-message-body` field to a string such as `New Message` for messages with a `<body/>` or `<encrypted/>` child.  But yes, this stuff should really be standardized.

[1]: https://xmpp.org/extensions/xep-0357.html#registrar-formtypes

**Comment by weiss** [comment]
One thing I'm unsure about is whether this should also be done for notifications triggered from MAM messages.  As per the [business rules suggested by @iNPUTmice][1], I would do this, but as ChatSecure currently doesn't support MAM, this would lead to bogus "New Message".

Does the current version of the `mod_cloud_notify` patch generate high-priority notifications for MAM messages?

[1]: https://mail.jabber.org/pipermail/standards/2016-February/030925.html

**Comment by tmolitor-stud-tu** [comment]
@weiss Yes it does obey those business rules completely. When there is no corresponding smacks session (either hibernated or alive) it does trigger push messages for MAM stanzas (and with Chris's patch MAM stanzas with `body` or `encrypted` elements are considered high-priority).

**Comment by gipfeljagd** [comment]
Since maybe a week, I receive four "New Message!" notifications, each time I receive a message from a chat partner. Chat partner gets "New Message!" notification only once. I have installed latest mod_cloud_notify.lua (with provided patch to get notification even if App is not running in background). Has this something to do with multiple "results" from pubsub.chatsecure.org?

`Sep 13 17:18:05 s2sin1c405e0	debug	Received[s2sin]: <iq id='86a7161e5956bec8263f96be45ed4a31ee1fa35524154545e1c559f7a96aebc2' type='result' to='somedomain.com' from='pubsub.chatsecure.org'>
Sep 13 17:18:05 stanzarouter	debug	Discarding iq from s2sin of type: result
Sep 13 17:18:05 s2sin1c405e0	debug	Received[s2sin]: <iq id='128dda856ce5b0e9181411f79cf846656dabae657f817d3c35a9f0e35ea1426e' type='result' to='somedomain.com' from='pubsub.chatsecure.org'>
Sep 13 17:18:05 stanzarouter	debug	Discarding iq from s2sin of type: result
Sep 13 17:18:05 s2sin1c405e0	debug	Received[s2sin]: <iq id='906dfbc7c007acefa20583f83e5b4c0104c8ab8784df1e5567b9bbcc9bbd6207' type='result' to='somedomain.com' from='pubsub.chatsecure.org'>
Sep 13 17:18:05 stanzarouter	debug	Discarding iq from s2sin of type: result
Sep 13 17:18:05 s2sin1c405e0	debug	Received[s2sin]: <iq id='0b5f8a3fdad18c0373463563c0cd4d95975d751e5c29245d287fc0d1087aa208' type='result' to='somedomain.com' from='pubsub.chatsecure.org'>
Sep 13 17:18:05 stanzarouter	debug	Discarding iq from s2sin of type: result`

**Comment by chrisballinger** [comment]
If you have duplicate push registrations, Go to Settings -> Account Details
(i) button -> Server Features -> Push -> Reset

**Comment by Rafaelrico77** [comment]
Hey, I have a question related to that topic. I use Prosody 0.9 with the default mod_cloud_notify downloaded with Mercury. But all of my users with Chatsecure (iOS) dont get the "New Message" Notification, when the App is closed. As I saw in the logs, the messages got forwarded correctly to the pupsub.chatsecure.com Server (no warnings or errors). Do I have to use the patched versions of the module, which are linked above or is the official one already patched, and there is maybe an other fault in my configuration or something ?

**Comment by mimi89999** [comment]
You need to use the version I linked.

**Comment by jsavage86** [comment]
@chrisballinger the patched mod_cloud_notify is working great; Thanks for your work!

@mimi89999 Thanks for the links.

Question: Does this patch require background activity to be enabled still?

**Comment by robbywraith** [comment]
Also had success with the mod_cloud_notify patch. Thanks! Any way we can increase visibility of this patch until it makes its way into the main branch? Took me forever to stumble across this post.

**Comment by Rafaelrico77** [comment]
@mimi89999 Thank You for the Links, the patched Version works well for me too.

An other Question: When I am right, the notifications work only in private Chat. When I understand it right, there is no support for group chat notifications on client site at the moment. Will this be fixed in the next version (4.1.1) ? And when will this version arrive in the App Store ?

**Comment by weiss** [comment]
> When I understand it right, there is no support for group chat notifications on client site at the moment.

The [group chat procotol][1] currently in use wasn't designed to support this kind of thing.  But we can work around the issue by keeping the [stream management][2] session alive while the client is disconnected, as described in the [business rules suggested by @iNPUTmice][3] (and my [followup message][4]).  This way, disconnected clients will remain joined to MUC rooms and receive push notifications for group chat messages.  ejabberd's `mod_push_keepalive` implements those business rules, and I *think* the current version of Prosody's push module supports them as well.

[1]: https://xmpp.org/extensions/xep-0045.html
[2]: https://xmpp.org/extensions/xep-0198.html
[3]: https://mail.jabber.org/pipermail/standards/2016-February/030925.html
[4]: https://mail.jabber.org/pipermail/standards/2016-February/030926.html

**Comment by tmolitor-stud-tu** [comment]
@weiss yes, prosody's mod_cloud_notify does implement this business rules, too (I added support several month ago).

General problem regarding this smacks session workaround is:
- groupchat messages only trigger silent pushes, which aren't very reliable
- swiping away the app (maybe accidental) kills the smacks session and all pending groupchat messages are lost (or is the smacks state persistent across app restarts? @chrisballinger )
- smacks sessions have a timeout (standard: 5 minutes) but can be higher. even then they usually arent' longer than a couple of hours. Not opening the app for this long invalidates the smacks session, too (assuming silent pushes aren't delivered properly; in my tests silent pushes got dropped by apple after about 30 pushes or after ~2 hours when the app was not opened in between)

**Conclusion:** Groupchats aren't very reliable. And to be honest it's not @chrisballinger's fault here. Apple is making it _very_ hard for developers to deliver a good push experience without sending the messages through Apples's servers (which would be a privacy nightmare).

**Sidenote:** WhatsApp is allowed to deliver high priority silent pushes because it implements VoIP calling (the infamous and rarely working WhatsApp calling feature). If Chatsecure also supported such a feature, it would be allowed to use those high priority pushes as well, which is why I have this VoIP stuff for Chatsecure on my ToDo-list (but I don't have time to implement it :( )
This VoIP stuff would even start the app on incoming messages when it was swiped away (force closed).

**Comment by weiss** [comment]
> groupchat messages only trigger silent pushes

You mean because the patched `mod_cloud_notify` doesn't mark groupchat messages as "high priority"?  If so, shouldn't this be changed?

> swiping away the app (maybe accidental) kills the smacks session and all pending groupchat messages are lost

The solution to this is implementing MUC MAM.

> smacks sessions have a timeout (standard: 5 minutes) but can be higher. even then they usually arent' longer than a couple of hours.

I [suggested][1] to keep the session alive until at least one push notification was generated.

[1]: https://mail.jabber.org/pipermail/standards/2016-February/030926.html

**Comment by tmolitor-stud-tu** [comment]
@weiss I didn't say these problems could not be somehow minimized (e.g. implement MAM), but currently these are the problems users will face and I just wanted to document them here :)

Regarding your suggestion to keep the session alive: I'll implement this in prosody. Does ejabberd do something similar?

**Comment by weiss** [comment]
> I didn't say these problems could not be somehow minimized (e.g. implement MAM), but currently these are the problems users will face and I just wanted to document them here :)

Lack of MUC MAM support should be somewhat compensated by MUC services sending some backlog when you join.  But yes, this may be incomplete (or nonexistent), of course.  The other two issues you mentioned should be solvable (or are already solved) by trivial changes to the server-side.

> Regarding your suggestion to keep the session alive: I'll implement this in prosody. Does ejabberd do something similar?

Yes, currently a separate module called `mod_push_keepalive` keeps the stream management session open for 24 hours (by default) if no push notification is generated during that time (the original timeout is restored on the first notification).  Before the session times out, a (silent) push notification is generated.

**Comment by jsavage86** [comment]
Not sure if this is the best place to do this but according to my testing, the patched mod_cloud_notify does not require background app refresh to be enabled. Hope this helps someone.

Also, @chrisballinger, because I have disabled background app refresh, it gives the red "X" under Server Info's Push Registration. If this patch becomes mainstream this part will need to be updated. Thanks for your work.

**Comment by andreygursky** [comment]
Has been https://modules.prosody.im/mod_cloud_notify.html already updated or where I can find a recent version dealing with this issue on iOS?

**Comment by chrisballinger** [comment]
@andreygursky No you will still need my patch. They don't want to merge it.

**Comment by andreygursky** [comment]
@chrisballinger, is the latest patched version available here: https://bitbucket.org/chrisballinger/prosody-modules/src/12b4387c0991eb24a15677b8bf59184ed880fa40/mod_cloud_notify/mod_cloud_notify.lua?at=default ?

I'm reading the success stories here and still can't believe that ChatSecure would become at last as usable as the proprietary software (since my communication partner blames me about no problems with whatsapp&Co and always working around free and open source on it's iOS). The only problem remaining to try it out is a server with the working mod_cloud_notify. That's why it would be great, if it goes upstream (it is already not so easy to convince a free server admin to pick up some module from hg tip, but applying a patch...). Were there any complains and modification requests or should one just ping upstream to take a look?

**Comment by chrisballinger** [comment]
@andreygursky No this is the latest: https://github.com/chrisballinger/prosody-modules/blob/xep_0357_priority/mod_cloud_notify/mod_cloud_notify.lua

The main thing is that this involves a change to the XMPP spec for XEP-0357, so beyond merging into Prosody it needs to get standardized for Ejabberd etc, and getting consensus on that is difficult.

**Comment by ageru** [comment]
Just had a brain fart: would the Prosody team maybe accept to include this altered version of mod_cloud_notify in [their community repo](https://hg.prosody.im/prosody-modules/) as a module on its own (mod_cloud_notify_ios_hack or something along those lines) ? Since this alternate module might be needed for a while, it would at least make discoverability and maintenance less difficult for server administrators (although I acknowledge it might be a pain to maintain outside of github for you Chris)

Anyway, it's just a thought, I leave the smell test to you ;)

**Comment by chrisballinger** [comment]
It should probably just be merged into the regular mod_cloud_notify because they already have some references to my patch in there.

**Comment by andreygursky** [comment]
@chrisballinger, thanks for pointing to the repo! So these couples of lines should make a true wonder:
```diff
--- prosody-modules.hg/mod_cloud_notify/mod_cloud_notify.lua    2017-10-17 19:39:18.970486913 +0200
+++ prosody-modules.git/mod_cloud_notify/mod_cloud_notify.lua   2017-11-04 20:02:15.516762311 +0100
@@ -14,6 +14,7 @@ local hashes = require"util.hashes";
 local xmlns_push = "urn:xmpp:push:0";
 
 -- configuration
+local include_priority = module:get_option_boolean("push_notification_with_priority", true);
 local include_body = module:get_option_boolean("push_notification_with_body", false);
 local include_sender = module:get_option_boolean("push_notification_with_sender", false);
 local max_push_errors = module:get_option_number("push_max_errors", 50);
@@ -270,6 +271,13 @@ local function handle_notify_request(sta
                        if stanza and include_body then
                                form_data["last-message-body"] = stanza:get_child_text("body");
                        end
+                       if stanza and include_priority then
+                               if stanza:get_child("body") or stanza:get_child("encrypted", "eu.siacs.conversations.axolotl") then
+                                       form_data["last-message-priority"] = "high"
+                               else
+                                       form_data["last-message-priority"] = "low"
+                               end
+                       end
                        push_publish:add_child(push_form:form(form_data));
                        if stanza and push_info.include_payload == "stripped" then
                                push_publish:tag("payload", { type = "stripped" })
```

**Comment by ThomasLeister** [comment]
Me and many other XMPP server admins I know would very much appreciate if this patch was also merged onto the official mod_cloud_notify version on modules.prosody.im! If someone has skills and time to maintain this thing - please go for it! :-)

**Comment by tmolitor-stud-tu** [comment]
@ThomasLeister I won't merge the patch because the `last-message-priority` field isn't specified in XEP-0357.

**But:** if the `last-message-body` field is set, Rubdub (the appserver used by ChatSecure) interpretes the push as "high priority", too.
I recently talked to @weiss (ejabberd developer) and we agreed to add a new config option to include a dummy message in `last-message-body` (something like "New message").

This way we won't break the standard, but still give you the functionality you need.

**Comment by dholl** [comment]
For anyone else still wrestling with push notifications even after implementing all the fine wisdom in this thread, also check for cipher suite compatibility issues noted in #822, such as ECDHE curves ```prime256v1``` vs ```secp384r1```...

**Comment by ageru** [comment]
Thank you for reporting this here @dholl . I still seem to be missing one piece of the puzzle, but whatever my problem is, now this curve/cipher issue is not it any more :)

**Comment by tmolitor-stud-tu** [comment]
FYI: I just pushed my proposed changes to modules.prosody.im over here: https://hg.prosody.im/prosody-modules/rev/df86ce6bb0b4

(the one I talked about over here: https://github.com/ChatSecure/ChatSecure-iOS/issues/770#issuecomment-368275301)

**Comment by ageru** [comment]
Hi @tmolitor-stud-tu and everyone,

So just to be sure I get everything right...

Currently, as a Prosody server admin, if I want my Chatsecure users to get notifications of new messages, I need to:
- Check for cipher suite compatibility, as per #822, and adapt cipher settings accordingly
- Install the latest version of mod_cloud_notify
- Insert a value for "push_notification_important_body", or is it not mandatory?
- Have MAM activated?
- ... (anything else I'm forgetting?)
- PROFIT?

Thanks again for all your contributions on this issue.

**Comment by tmolitor-stud-tu** [comment]
@ageru Yes, `push_notification_important_body` has to be set to some text like "New Message".
But to be honest it can be anything because ChatSecure's AppServer strips the text off when sending the notification via APNS.
It's just needed for the AppServer to know the push is a high priority one.

**Comment by chrisballinger** [comment]
@ageru @tmolitor-stud-tu Please don't send us message content, make sure to replace it with something static.

**Comment by ldvc** [comment]
In https://github.com/ChatSecure/ChatSecure-iOS/issues/770#issuecomment-368275301:
>I recently talked to @weiss (ejabberd developer) and we agreed to add a new config option to include a dummy message in last-message-body (something like "New message").

@weiss is this "dummy message" still planned for ejabberd?

**Comment by weiss** [comment]
> @weiss is this "dummy message" still planned for ejabberd?

Yes, I'll add this soon™.

**Comment by Lyokovic** [comment]
Hi,

Just a comment to say that after updating my prosody modules and adding the `push_notification_important_body` option, push notifications now work for me.

Thanks all for your work!

**Comment by ageru** [comment]
I was despairing a bit after Lyokovic's message, since I was still getting the same "cancel" message as before.

So I tested a few things this evening, and it seems what did the trick for me was to activate S2S, which had been off until then. Does it means that S2S is required for the connection to pubsub.chatsecure.org to happen? Is that expected behaviour? This is a complete candid question,  as I was expecting S2S to be for pure messaging only, not in this context.

Here is the config that was successful:
- ChatSecure 4.3
- Conversation 2.0.0 on the other side of the conversation
- Prosody 0.10.0
- All that [cipher jazz](https://github.com/ChatSecure/ChatSecure-iOS/issues/822#issuecomment-371726309) set properly
- main modules enabled: mam (internal), smacks, smacks_offline, pep_vcard_avatar, csi, filter_chatstates, throttle_presence, cloud_notify...
- ...and S2S enabled as of a few minutes ago.

I'll keep testing in the next few day. Please do chip in with your insights and your own experience.

Fingers crossed I'm closing this one soon...

**Comment by weiss** [comment]
> Does it means that S2S is required for the connection to pubsub.chatsecure.org to happen? Is that expected behaviour?

Yes.

**Comment by weiss** [comment]
@danielreuther 
> > > @weiss is this "dummy message" still planned for ejabberd?
> >
> > Yes, I'll add this soon™.
>
> Sounds great. In case you have a development branch that contains it, I'd be happy to try it out.

[It][1]'s in the `master` branch, now.  In order to enable it:

```yml
    modules:
      mod_push:
        include_body: "New message"
```

[1]: https://github.com/processone/ejabberd/commit/de7dc4affa6c6996bed838b5d4b0b53511962ef0

**Comment by antonyablonsky** [comment]
Hi everyone! Is there way to make s2s connection between ejabberd server and pubsub.chatsecure.org if server was not configured with DNS records and domain name like "example.com"? it just lives on my static ip and I connect straightforward, like user@ip_adress. Thanks

**Comment by chrisballinger** [comment]
Nope sorry. It requires a valid SSL certificate. You can get domains for
pretty cheap these days like $2/year, and if you use Let's Encrypt you can
get free SSL certificates.

**Comment by Echolon** [comment]
Hi,

just started an EtherCalc Pad to collect information about your and others experience about push. Feel free to add your data or additional columns (suggest only to add columns where the information can be given from a normal user)

https://ethercalc.org/29iyo1nunqdp

Cheers

**Comment by ageru** [comment]
For the past few months, I have been running Prosody 0.10.2, and my users have been running ChatSecure 4.3 and latest. Right now:
- I don't get anything out of the ordinary in the logs
- iOS users mostly get push messages.

I says "mostly" because there are still 2 instances when the ChatSecure users don't get messages
- One is with group chat, but that's an issue with MAM on Prosody, and it impacts Android also.
- due to iOS closing the app after some time, which is kind of an expected issue at this stage of ChatSecure's development.

But those are outside the scope of my initial issue, wich I consider solved, so I'm closing tis bug. Thanks everyone for participating!

Here is how my server is set up to allow for push, in case this is useful to anyone:
- Running prosody 0.10.2, but should mostly work on the 0.9 branch
- S2S enabled ("Server-to-server" refers to any communication between servers, not just the transmission of messages between users on different servers. So it needs to be on for push to work, as push involves server-to-server communication.
- dialback doesn't need to be enabled
- [TLS Ciphers set-up in a very specific way](https://github.com/ChatSecure/ChatSecure-iOS/issues/822#issuecomment-371726309), but hopefully this will be solved at some point.
- main modules enabled: mam (internal), smacks, smacks_offline, pep_vcard_avatar, csi, filter_chatstates, throttle_presence, cloud_notify
- users running Conversations, Xabber, ChatSecure.

**Comment by Echolon** [comment]
Better leave this issue open

**Comment by STPKITT** [comment]
> iOS users mostly get push messages.
I says "mostly" because there are still 2 instances when the ChatSecure users don't get messages

> due to iOS closing the app after some time, which is kind of an expected issue at this stage of ChatSecure's development.

But isn't iOS closing the app the issue what push was actually implemented for in the first place? That's *the* main tripping hazard the ordinary end user faces & ultimately the deal breaker which keeps them on proprietary instant messaging services.
So if this issue stays closed, which issue # is the relevant one for this remaining problem?

**Comment by robinde** [comment]
Trawling through the various issues about push notifications it's not clear what is up to date and what is not. What's the current requirement for a working prosody setup with push notifications?
* latest version of `mod_cloud_notify` is clearly needed
* do we still need the [very specific cipher setup](https://github.com/ChatSecure/ChatSecure-iOS/issues/822#issuecomment-371726309)
* what about needing to set a random value for `push_notification_important_body = "New Message!"`
* other required modules? `pubsub`, `pinger` ?

**Comment by licaon-kter** [comment]
> latest version of mod_cloud_notify is clearly needed

yes

> what about needing to set a random value for push_notification_important_body = "New Message!"

Is already included: https://hg.prosody.im/prosody-modules/file/tip/mod_cloud_notify/mod_cloud_notify.lua#l25
